File: SPOJ/TOUR.java

import java.io.*;
import java.util.*;
public class TOUR {
	public static void main(String[] args)throws Throwable {
		MyScanner sc=new MyScanner();
		PrintWriter pw=new PrintWriter(System.out);
		
		int t=sc.nextInt();
		while(t-->0){
			n=sc.nextInt();
			adj=new ArrayList [n];
			vis=new int [n];
			Arrays.fill(vis, -1);
			for(int i=0;i<n;i++)
				adj[i]=new ArrayList<Integer>();
			for(int i=0;i<n;i++){
				int c=sc.nextInt();
				while(c-->0){
					int v=sc.nextInt()-1;
					adj[v].add(i);
				}
			}
			reach=new boolean [n][n];
			for(int i=0;i<n;i++){
				cur=i;
				dfs(i);
			}
			int ans=0;
			for(int i=0;i<n;i++){
				boolean ok=true;
				for(int j=0;j<n;j++)
					if(!reach[i][j])
						ok=false;
				if(ok)
					ans++;
			}
			
			pw.println(ans);
		}
		
		pw.flush();
		pw.close();
	}
	static int n,cur;
	static int [] vis;
	static boolean [][] reach;
	static ArrayList<Integer> [] adj;
	
	static void dfs(int u){
		reach[cur][u]=true;
		vis[u]=cur;
		for(int v : adj[u])
			if(vis[v]!=cur)
				dfs(v);
	}
	
	static class MyScanner {
		BufferedReader br;
		StringTokenizer st;
		public MyScanner() {br = new BufferedReader(new InputStreamReader(System.in));}
		String next() {while (st == null || !st.hasMoreElements()) {
			try {st = new StringTokenizer(br.readLine());}
			catch (IOException e) {e.printStackTrace();}}
		return st.nextToken();}
		int nextInt() {return Integer.parseInt(next());}
		long nextLong() {return Long.parseLong(next());}
		double nextDouble() {return Double.parseDouble(next());}
		String nextLine(){String str = "";
		try {str = br.readLine();}
		catch (IOException e) {e.printStackTrace();}
		return str;}
	}
}

File: JAVA/SPOJ/TOUR.cpp
#include <bits/stdc++.h>
using namespace std;
 
const int N=1000;
int cur;
vector<int> adj[N];
bool reach [N][N];
int vis[N];
void dfs(int u){
	reach[cur][u]=true;
	vis[u]=cur;
	for(int i=0;i<adj[u].size();i++){
		int v=adj[u][i];
		if(vis[v]!=cur)
			dfs(v);
	}
}
 
 
int main(){
	int t;
	scanf("%d",&t);
	while(t--){
		int n;
		scanf("%d",&n);
		memset(vis, -1,sizeof vis);
		for(int i=0;i<n;i++)
			adj[i].clear();
		for(int i=0;i<n;i++){
			int c;
			scanf("%d",&c);
			while(c--){
				int v;
				scanf("%d",&v);
				v--;
				adj[v].push_back(i);
			}
		}
		for(int i=0;i<n;i++)
			for(int j=0;j<n;j++)
				reach[i][j]=false;
		for(int i=0;i<n;i++){
			cur=i;
			dfs(i);
		}
		int ans=0;
		for(int i=0;i<n;i++){
			bool ok=true;
			for(int j=0;j<n;j++)
				if(!reach[i][j])
					ok=false;
			if(ok)
				ans++;
		}
 
		printf("%d\n",ans);
	}
}
 